python_modules/dagster/dagster/_core/execution/context/system.py

```diff
@@ -888,35 +903,80 @@ def fetch_external_input_asset_records(self) -> None:
                 if key not in all_dep_keys and key not in output_keys:
                     all_dep_keys.append(key)
 
-        self._input_asset_records = {}
+        self._input_asset_version_info = {}
         for key in all_dep_keys:
-            self._fetch_input_asset_record(key)
-        self._is_external_input_asset_records_loaded = True
+            self._fetch_input_asset_version_info(key)
+        self._is_external_input_asset_version_info_loaded = True
 
-    def _fetch_input_asset_record(self, key: AssetKey, retries: int = 0) -> None:
+    def _fetch_input_asset_version_info(self, key: AssetKey) -> None:
         from dagster._core.definitions.data_version import (
             extract_data_version_from_entry,
         )
 
-        event = self.instance.get_latest_data_version_record(key)
-        if key in self._data_version_cache and retries <= 5:
-            event_data_version = (
-                None if event is None else extract_data_version_from_entry(event.event_log_entry)
-            )
-            if event_data_version == self._data_version_cache[key]:
-                self._input_asset_records[key] = event
+        event = self._get_input_asset_event(key)
+        if event is None:
+            self._input_asset_version_info[key] = None
+        else:
+            storage_id = event.storage_id
+            # Input name will be none if this is an internal dep
+            input_name = self.job_def.asset_layer.input_for_asset_key(self.node_handle, key)
+            if input_name and self.has_asset_partitions_for_input(input_name):
```

Comment:
my main concern here is just situations in which we have a `<huge number of partitions>` -> `<unpartitioned asset>` connection, which would require us to fire off a massive query in order to get the latest data versions per each one of those partitions.

it's definitely possible that this is kinda fine in practice, but for this PR I think we can still claim an objective improvement over status quo if we filter out cases where there's an AllPartitionMapping. Essentially, `if input_name and self.has_asset_partitions_for_input and not self._has_all_partition_mapping(input): ...` where has_all_partition_mapping basically just grabs the partition_mapping in the same way as asset_partitions_subset_for_input and makes sure it's not an AllPartitionMapping()

This change is less principled but I think shaves off a bunch of risk from this implementation and is easy to remove at a later point.